listenbrainz/webserver/static/js/src/CBReviewModal.tsx

```diff
@@ -0,0 +1,710 @@
+  getGroupMBIDFromRelease = async (mbid: string): Promise<string> => {
+    const { APIService } = this.context;
+
+    try {
+      const response = await APIService.lookupMBRelease(mbid);
+      return response["release-group"].id;
+    } catch (error) {
+      this.handleError(error, "Could fetch release group MBID");
+      return "";
+    }
+  };
+
+  getRecordingMBIDFromTrack = async (
+    mbid: string,
+    track_name: string
+  ): Promise<string> => {
+    const { APIService } = this.context;
+
+    try {
+      const response = await APIService.lookupMBReleaseFromTrack(mbid);
+      // MusicBrainz API returns multiple releases, medias, and tracks, so we need to
+      // search for the track with a name that matches the supplied track_name
+      const medias = response.releases[0].media;
+      const media = medias.find((res: any) => res.tracks);
+
+      if (media) {
+        const track = media.tracks.find(
+          (res: any) =>
+            res.recording.title.toLowerCase() === track_name.toLowerCase()
+        );
+        if (track) return track.recording.id;
+      }
+      return "";
+    } catch (error) {
+      this.handleError(error, "Could fetch recording MBID");
+      return "";
+    }
+  };
+
+  /* determine entity functions */
+  getAllEntities = async () => {
+    this.setState({ loading: true });
+
+    // get all three entities and then set the default entityToReview
+    this.getArtistEntity();
+    await this.getRecordingEntity();
+    await this.getReleaseGroupEntity();
+
+    this.setEntityToReview();
+    this.setState({ loading: false });
+  };
+
+  getRecordingEntity = async () => {
+    this.setState({ recordingEntity: null });
+
+    const { listen } = this.props;
+    const { additional_info } = listen.track_metadata;
+
+    let recording_mbid = "";
+
+    if (additional_info?.recording_mbid)
+      recording_mbid = additional_info?.recording_mbid;
+    // If listen doesn't contain recording_mbid attribute,
+    // search for it using the track mbid instead
+    else if (additional_info?.track_mbid)
+      recording_mbid = await this.getRecordingMBIDFromTrack(
+        additional_info?.track_mbid,
+        listen.track_metadata.track_name
+      );
+    else if (additional_info?.lastfm_track_mbid)
+      recording_mbid = await this.getRecordingMBIDFromTrack(
+        additional_info?.lastfm_track_mbid,
+        listen.track_metadata.track_name
+      );
+
+    // confirm that found mbid was valid
+    if (recording_mbid.length) {
+      const entity: ReviewableEntity = {
+        type: "recording",
+        mbid: recording_mbid,
+        name: listen.track_metadata.track_name,
+      };
+      this.setState({ recordingEntity: entity });
+    }
+  };
+
+  getArtistEntity = () => {
+    this.setState({ artistEntity: null });
+
+    const { listen } = this.props;
+    const { additional_info } = listen.track_metadata;
+    let artist_mbid;
+
+    if (additional_info?.lastfm_artist_mbid)
```

Comment:
Oh! I see its used here. My understanding from the chat logs was that lastfm supplied mbids are not trust worthy and hence should not be used. @mayhem thoughts?